test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusDevModeTest.java

```diff
@@ -110,6 +127,7 @@ public Object createTestInstance(TestInstanceFactoryContext factoryContext, Exte
 
     @Override
     public void beforeEach(ExtensionContext extensionContext) throws Exception {
+        rootLogger.addHandler(inMemoryLogHandler);
```

Comment:
I'm not too familiar with loggers and junit extensions, but isn't this going to leak in the testsuite?

Looks like it will keep adding new InMemoryLogHandler instances, and also it's not emptying each of them after usage. I suppose the `rootLogger` is a global static so this might get nasty?

Reply:
@geoand any opinion? I copied this from your `QuarkusProdModeTest`.

Reply:
In `QuarkusProdModeTest` this is done in `beforeAll`, is there any specific reason why it was added to `beforeEach` here?

Reply:
Having it in `beforeAll` might be slightly better, but would still leak no?

Reply:
Sorry I don't follow, what would leak?

Reply:
Maybe just add it in a static block, so you're sure it's done once across the whole classloader.

Then, I'd use an "afterEach" or "afterAll" to clear its contents after each unit of testing.

Reply:
> Sorry I don't follow, what would leak?

I suspect two leaks, although I haven't debugged this and I'm not too familiar with this code so I might be wrong:

1. The root logger is a global static instance, so each time we invoke `addHandler` we're adding more and more loghandlers.

2. each such logHandler will keep containing all messages, from since it was added to the root onwards - so also all text generated from other tests running later in the testsuite.

Reply:
@geoand I've put it where we start the dev mode, which is in "beforeEach", while the prod test is started in "beforeAll".

Reply:
I've now removed it in `afterEach`, should be enough no?

Reply:
That isn't really needed because each time a the extension is run (that is each time we write a `QuarkusDevModeTest` or `QuarkusProdModeTest`), the handler that is used for that test **replaces** the global static.

Reply:
> the handler that is used for that test replaces the global static.

But the `private static final Logger rootLogger;` isn't replaced is it? Or is each test running in its own classloader?

Reply:
The `rootLogger` is the same for all tests (they don't run in isolared classloaders), but because each test replaces it, it shouldn't be a problem.

There could be an issue when running `QuarkusDevModeTest` other tests inside the same test execution (read `surefire` execution) as in that case indeed the handler set from `QuarkusDevModeTest` will persist.
So the safest thing would indeed be to restore the original handlers.

Should just be a matter of calling and saving `getHandlers` before `addHandler` and then in the `afterAll`, restoring the handlers using `setHandlers` (after casting to `org.jboss.logmanager.Logger`).

Would you like me to do that @FroMage or do you want to include it here?

All in all, although it's unlikely to cause a problem, we should fix it and it's a very good catch @Sanne

Reply:
I'm still not understanding what you mean by "but because each test replaces it, it shouldn't be a problem" but it's ok, we can avoid blocking this PR just because I don't get it :)

With @FroMage having added the `afterEach` I'm happy.

Reply:
I mean that each new instance of QuarkusDevModeTest will provide a new handler thus replacing the old one.
But for sure the best solution is to get a hold of the initial handlers in beforeAll and then restore them in afterAll

Reply:
it's adding handlers, not replacing existing ones.

Reply:
I'll take care of this in a separate PR

Reply:
https://github.com/quarkusio/quarkus/pull/10262